src/shogun/base/AnyParameter.h

```diff
@@ -28,101 +28,85 @@ namespace shogun
+		/** Default constructor where all parameter properties are false
 		 */
-		AnyParameterProperties() : m_description(), m_mask_attribute(0x00000000)
-		{
-		}
-		/** legacy constructor */
+		AnyParameterProperties()
+		    : m_description("No description given"), m_attribute_mask(0)
+		{
+		}
+		/** Constructor
+		 * @param description parameter description
+		 * @param hyperparameter set to true for parameters that determine
+		 * how training is performed, e.g. regularisation parameters
+		 * @param gradient set to true for parameters required for gradient
+		 * updates
+		 * @param model set to true for parameters used in inference, e.g.
+		 * weights and bias
+		 * */
 		AnyParameterProperties(
 		    std::string description,
-		    EModelSelectionAvailability model_selection = MS_NOT_AVAILABLE,
+		    EModelSelectionAvailability hyperparameter = MS_NOT_AVAILABLE,
 		    EGradientAvailability gradient = GRADIENT_NOT_AVAILABLE,
-		    bool hyperparameter = false)
-		    : m_description(description), m_model_selection(model_selection),
+		    bool model = false)
+		    : m_description(description), m_model_selection(hyperparameter),
 		      m_gradient(gradient)
 		{
-			m_mask_attribute = 0x00000000;
-			if (model_selection)
-				m_mask_attribute |= MODEL_PARAM;
-			if (gradient)
-				m_mask_attribute |= GRADIENT_PARAM;
-			if (hyperparameter)
-				m_mask_attribute |= HYPERPARAMETER;
+			m_attribute_mask = (hyperparameter << 0 & HYPER) |
+			                   (gradient << 1 & GRADIENT) |
+			                   (model << 2 & MODEL);
 		}
-		/** mask constructor */
-		AnyParameterProperties(std::string description, int32_t mask_attribute)
+		/** Mask constructor
+		 * @param description parameter description
+		 * @param attribute_mask mask encoding parameter properties
+		 * */
+		AnyParameterProperties(std::string description, int32_t attribute_mask)
 		    : m_description(description)
 		{
-			m_mask_attribute = mask_attribute;
+			m_attribute_mask = attribute_mask;
 		}
-		/** copy contructor */
+		/** Copy contructor */
 		AnyParameterProperties(const AnyParameterProperties& other)
 		    : m_description(other.m_description),
 		      m_model_selection(other.m_model_selection),
 		      m_gradient(other.m_gradient),
-		      m_mask_attribute(other.m_mask_attribute)
+		      m_attribute_mask(other.m_attribute_mask)
 		{
 		}
-		/** description getter */
-		std::string get_description() const
+		const std::string& get_description() const
 		{
 			return m_description;
 		}
-		/** model selection flag getter */
 		EModelSelectionAvailability get_model_selection() const
 		{
-			EModelSelectionAvailability return_val;
-			if (m_mask_attribute & HYPERPARAMETER)
-				return_val = EModelSelectionAvailability::MS_AVAILABLE;
-			else
-				return_val = EModelSelectionAvailability::MS_NOT_AVAILABLE;
-			return return_val;
+			return static_cast<EModelSelectionAvailability>(
+			    (m_attribute_mask & HYPER) > 0);
```

Comment:
and here I was thinking
`m_attribute_mask & (1<<HYPER)`

Reply:
wouldn't we want != 0. Also, this is not very explicit ... I know sergey suggested, but I need to think a bit here to understand this so not sure that is the best solution

Reply:
I thought about that, but then I thought it was cleaner to have the constructor "know" at what position the bit shift is required? Not sure if that makes sense?

Reply:
not sure :) haha
we can discuss in irc